### PR TITLE
YAML loading and dumping improvements

### PR DESCRIPTION
**Pull request by rhansen**
### What does this PR do?

Introduces a new `yaml_compatibility` option that controls the behavior of YAML loading/dumping.  If a user wants to revert to the YAML behavior from a previous version of Salt, or preview upcoming changes to YAML behavior, they can set it to the desired version of Salt.  For example, setting `yaml_compatibility` to 3006 will pin the current behavior, and setting it to 3007 allows the user to preview the behavior that will become the default in Salt v3007.

In addition, the following behavior changes take effect immediately:

  * A warning is logged saying that the default YAML behavior will change in Salt v3007.
  * Passing a non-`dict` class to the `salt.utils.yaml.SaltYamlSafeLoader` constructor no longer causes all future `!!omap` nodes to raise an exception when loaded.
  * Loading a sequence node tagged with `!!python/tuple` is now supported, and produces a Python `tuple` object.  Before, such nodes would raise an exception.  (This change does not break backwards compatibility, which is why it takes effect immediately.)

After v3006 is released (`salt.version.SaltVersionsInfo.SULFUR.released` is changed to `True`), the following behavior changes will automatically take effect.  These can be previewed now by setting `yaml_compatibility` to 3007.

  * `salt.utils.yaml.dump()` will default to `salt.utils.yaml.OrderedDumper` rather than PyYAML's `yaml.Dumper` class. (I believe this was the original intention, and now `dump()`'s behavior is symmetric with `safe_dump()`).
  * Fix `salt.utils.yaml.IndentedSafeOrderedDumper` list indentation (was broken only when `yaml.CSafeDumper` exists).
  * Loading an explicitly tagged YAML `!!timestamp` node will produces a Python `datetime.datetime` object (currently they are always loaded as strings).
  * Fix loading of YAML `!!omap` nodes (currently they are sometimes incorrectly parsed as mapping nodes rather than sequence nodes containing mapping nodes).
  * Loading a YAML `!!omap` node will always produce a `collections.OrderedDict` object.  Currently, it sometimes produces a list of (key, value) tuples and sometimes raises an exception.
  * Dumping a Python `datetime.datetime` object to YAML now explicitly tags the node with `!!timestamp`.
  * Dumping a Python `collections.OrderedDict` object to YAML now consistently produces a YAML `!!omap` node, which is a sequence of single-valued mappings.  Before, the behavior varied depending on the specific object type (`collections.OrderedDict` vs. `salt.utils.odict.OrderedDict`) and the dump function (`salt.utils.yaml.dump()` vs. `salt.utils.yaml.safe_dump()`).
  * Dumping a Python `tuple` object to YAML now consistently produces a YAML sequence node tagged with `!!python/tuple`.  Before, only `salt.utils.yaml.dump()` would tag the node with `!!python/tuple`; `salt.utils.yaml.safe_dump()` did not.
  * Dumping an object of unsupported type using a safe dumper will raise an exception.  Currently it produces a `NULL` node.

Clean-ups and code health:
  * Convert `salt.utils.yaml` unit tests to pytest.
  * Add an integration test that demonstrates that #12161 is already fixed (probably when Salt dropped support for Python <3.6).
  * Factor out duplicate code.
  * Improve code readability.

### What issues does this PR fix or reference?

#12161 (can be closed even without this PR because the bug no longer exists, although this PR adds a regression test in the form of an integration test)

### Behavior Changes

  * `IndentedSafeOrderedDumper`:
      * Before (assuming `yaml.CSafeDumper` exists):

        ```pycon
        >>> import salt.utils.yaml as y
        >>> print(y.dump({"foo": ["bar"]}, Dumper=y.IndentedSafeOrderedDumper, default_flow_style=False))
        foo:
        - bar
        
        ```

      * After:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> print(y.dump({"foo": ["bar"]}, Dumper=y.IndentedSafeOrderedDumper, default_flow_style=False))
        foo:
          - bar
        
        ```

        This fix required a switch to the pure Python `yaml.SafeDumper` class (for `IndentedSafeOrderedDumper` only), which will be slower than `yaml.CSafeDumper`.

  * Loading YAML `!!omap` nodes to Python:
      * Before:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> import collections
        >>> y.load("!!omap [{foo: bar}, {baz: bif}]")
        [('foo', 'bar'), ('baz', 'bif')]
        >>> y.SaltYamlSafeLoader("", dictclass=collections.OrderedDict)  # created for side-effect only
        <salt.utils.yamlloader.SaltYamlSafeLoader object at 0xc31a10>
        >>> y.load("!!omap [{foo: bar}, {baz: bif}]")  # exact same as before
        Traceback (most recent call last):
          File "<stdin>", line 1, in <module>
          File "salt/utils/yamlloader.py", line 159, in load
            return yaml.load(stream, Loader=Loader)
          File "venv/lib/python3.8/site-packages/yaml/__init__.py", line 81, in load
            return loader.get_single_data()
          File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
            return self.construct_document(node)
          File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 60, in construct_document
            for dummy in generator:
          File "salt/utils/yamlloader.py", line 45, in construct_yaml_map
            value = self.construct_mapping(node)
          File "salt/utils/yamlloader.py", line 56, in construct_mapping
            raise ConstructorError(
        yaml.constructor.ConstructorError: expected a mapping node, but found sequence
          in "<unicode string>", line 1, column 1
        ```

      * After:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> import collections
        >>> y.load("!!omap [{foo: bar}, {baz: bif}]")
        OrderedDict([('foo', 'bar'), ('baz', 'bif')])
        >>> y.SaltYamlSafeLoader("!!omap [{foo: bar}, {baz: bif}]", dictclass=collections.OrderedDict).get_single_data()
        OrderedDict([('foo', 'bar'), ('baz', 'bif')])
        >>> y.load("!!omap [{foo: bar}, {baz: bif}]")
        OrderedDict([('foo', 'bar'), ('baz', 'bif')])
        ```

  * Loading YAML `!!timestamp` nodes to Python:
      * Before:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> y.load("!!timestamp 2001-12-15T02:59:43.1Z")
        '2001-12-15T02:59:43.1Z'
        >>> y.load("2001-12-15T02:59:43.1Z")  # implicit timestamp
        '2001-12-15T02:59:43.1Z'
        ```

      * After:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> y.load("!!timestamp 2001-12-15T02:59:43.1Z")
        datetime.datetime(2001, 12, 15, 2, 59, 43, 100000, tzinfo=datetime.timezone.utc)
        >>> y.load("2001-12-15T02:59:43.1Z")  # untagged is still a string
        '2001-12-15T02:59:43.1Z'
        ```

  * Loading YAML `!!python/tuple` nodes to Python:
      * Before:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> y.load("!!python/tuple [foo, bar]")
        Traceback (most recent call last):
          File "<stdin>", line 1, in <module>
          File "salt/utils/yamlloader.py", line 159, in load
            return yaml.load(stream, Loader=Loader)
          File "venv/lib/python3.8/site-packages/yaml/__init__.py", line 81, in load
            return loader.get_single_data()
          File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
            return self.construct_document(node)
          File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 55, in construct_document
            data = self.construct_object(node)
          File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 100, in construct_object
            data = constructor(self, node)
          File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 427, in construct_undefined
            raise ConstructorError(None, None,
        yaml.constructor.ConstructorError: could not determine a constructor for the tag 'tag:yaml.org,2002:python/tuple'
          in "<unicode string>", line 1, column 1
        ```

      * After:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> y.load("!!python/tuple [foo, bar]")
        ('foo', 'bar')
        ```

  * Dumping Python `datetime.datetime` objects to YAML:
      * Before:

        ```pycon
        >>> import datetime
        >>> import salt.utils.yaml as y
        >>> print(y.dump(datetime.datetime(2022, 10, 21, 18, 16, 3, 100000, tzinfo=datetime.timezone(datetime.timedelta(hours=-4)))))
        2022-10-21 18:16:03.100000-04:00
        ```

      * After:

        ```pycon
        >>> import datetime
        >>> import salt.utils.yaml as y
        >>> print(y.dump(datetime.datetime(2022, 10, 21, 18, 16, 3, 100000, tzinfo=datetime.timezone(datetime.timedelta(hours=-4)))))
        !!timestamp 2022-10-21 18:16:03.100000-04:00
        ```

  * Dumping Python `OrderedDict` objects to YAML:
      * Before:

        ```pycon
        >>> from salt.utils.odict import OrderedDict
        >>> import salt.utils.yaml as y
        >>> import collections
        >>> print(y.dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
        !!python/object/apply:salt.utils.odict.OrderedDict
        dictitems:
          foo: bar
        
        >>> print(y.safe_dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
        foo: bar
        
        >>> print(y.dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
        !!python/object/apply:collections.OrderedDict
        - - - foo
            - bar
        
        >>> print(y.safe_dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
        NULL
        
        ```

      * After:

        ```pycon
        >>> from salt.utils.odict import OrderedDict
        >>> import salt.utils.yaml as y
        >>> import collections
        >>> print(y.dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
        !!omap
        - foo: bar
        
        >>> print(y.safe_dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
        !!omap
        - foo: bar
        
        >>> print(y.dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
        !!omap
        - foo: bar
        
        >>> print(y.safe_dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
        !!omap
        - foo: bar
        
        ```

  * Dumping Python `tuple` objects to YAML:
      * Before:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> print(y.dump(("foo", "bar"), default_flow_style=False))
        !!python/tuple
        - foo
        - bar
        
        >>> print(y.safe_dump(("foo", "bar"), default_flow_style=False))
        - foo
        - bar
        
        ```

      * After:

        ```pycon
        >>> import salt.utils.yaml as y
        >>> print(y.dump(("foo", "bar"), default_flow_style=False))
        !!python/tuple
        - foo
        - bar
        
        >>> print(y.safe_dump(("foo", "bar"), default_flow_style=False))
        !!python/tuple
        - foo
        - bar
        
        ```

  * Dumping unsupported objects:
      * Behavior before:
    
        ```pycon
        >>> import salt.utils.yaml as y
        >>> y.safe_dump(type("GeneratedObjectSubclass", (), {}))
        'NULL\n'
        ```
    
      * Behavior after:
    
        ```pycon
        >>> import salt.utils.yaml as y
        >>> y.safe_dump(type("GeneratedObjectSubclass", (), {}))
        Traceback (most recent call last):
          File "<stdin>", line 1, in <module>
          File "salt/utils/yamldumper.py", line 270, in safe_dump
            return dump(data, stream, Dumper=SafeOrderedDumper, **kwargs)
          File "salt/utils/yamldumper.py", line 261, in dump
            return yaml.dump(data, stream, **kwargs)
          File "venv/lib/python3.8/site-packages/yaml/__init__.py", line 253, in dump
            return dump_all([data], stream, Dumper=Dumper, **kwds)
          File "venv/lib/python3.8/site-packages/yaml/__init__.py", line 241, in dump_all
            dumper.represent(data)
          File "venv/lib/python3.8/site-packages/yaml/representer.py", line 27, in represent
            node = self.represent_data(data)
          File "venv/lib/python3.8/site-packages/yaml/representer.py", line 58, in represent_data
            node = self.yaml_representers[None](self, data)
          File "venv/lib/python3.8/site-packages/yaml/representer.py", line 231, in represent_undefined
            raise RepresenterError("cannot represent an object", data)
        yaml.representer.RepresenterError: ('cannot represent an object', <class '__main__.GeneratedObjectSubclass'>)
        ```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

### Tasks
  * [x] write a changelog entry
  * [x] figure out where to document these changes
  * [x] Add a feature flag to enable this new behavior.
  * [x] Get all the tests to pass.
  * [ ] Get #63073 merged

### CODE REVIEW

**Comment by rhansen** [comment]
This PR is marked as draft until I complete the following tasks:
  * [ ] write a changelog entry
  * [ ] figure out where to document these changes
  * [ ] ensure that all tests pass

**Comment by rhansen** [comment]
I believe this is ready to be reviewed now.

File: salt/utils/yamldumper.py [hunk 1]
```diff
@@ -108,12 +126,20 @@ def dump(data, stream=None, **kwargs):
...
+        ``yaml.Dumper``.
+
     Helper that wraps yaml.dump and ensures that we encode unicode strings
     unless explicitly told not to.
     """
-    if "allow_unicode" not in kwargs:
-        kwargs["allow_unicode"] = True
-    kwargs.setdefault("default_flow_style", None)
+    kwargs = {
+        "allow_unicode": True,
```
Comment (MKLeb) [inline salt/utils/yamldumper.py:138]:
```suggestion
        "allow_unicode": kwargs.get("allow_unicode", True),
```

Just so this doesn't become a breaking change?

File: tests/pytests/unit/utils/jinja/test_custom_extensions.py [hunk 2]
```diff
@@ -131,10 +132,13 @@ def test_serialize_yaml():
         "baz": [1, 2, 3],
         "qux": 2.0,
         "spam": OrderedDict([("foo", OrderedDict([("bar", "baz"), ("qux", 42)]))]),
+        "tuple": ("foo", "bar"),
     }
     env = Environment(extensions=[SerializerExtension])
     rendered = env.from_string("{{ dataset|yaml }}").render(dataset=dataset)
     assert dataset == salt.utils.yaml.safe_load(rendered)
+    assert isinstance(dataset["spam"], collections.OrderedDict)
+    assert isinstance(dataset["spam"]["foo"], collections.OrderedDict)
```
Comment (MKLeb) [inline tests/pytests/unit/utils/jinja/test_custom_extensions.py:141]: I'm going to trigger a full test run on this PR, simply to find out if there are any other tests that may fail due to these changes.

**Comment by MKLeb** [comment]
re-run full all

**Comment by waynew** [comment]
FWIW I'm on board with this idea! I don't yet know if I'm on board with the implementation - as @MKLeb implied with the full test run, this is a pretty intensive change with (potentially) broad repercussions... hopefully all of them good! But I'm sure someone out there is relying on some backwards funky behavior :upside_down_face: 

I would *love* if we could bridge the existing behavior and this with like... a feature flag and deprecation warning, like...

```
if not __opts__['use_up_to_date_yaml_dumping']:
    raise DeprecationWarning('You are currently using the legacy YAML loading/dumping approach. This will work as exists until 2024 at the earliest. To use the updated YAML behavior, set `use_up_to_date_yaml_dumping: True` in your config files.")
    do_old_stuff()
else:
    do_this()
```

That way folks could opt-in to the new behavior, and if they have a bunch of problems then they can file some bugs and still have a better fallback than just downgrading their salt installation.

In any case, I'm sure a lot of us would feel better with an in-depth code review!

File: salt/utils/yamldumper.py (hunk 1 above)
Comment (rhansen) [reply salt/utils/yamldumper.py:138]: No need.  If `"allow_unicode"` is already in `kwargs` then the `**kwargs` line below will cause its value to replace this value.

**Comment by rhansen** [comment]
> I would _love_ if we could bridge the existing behavior and this with like... a feature flag and deprecation warning

I can do that.  One thing:  There is already an option for legacy YAML behavior: [`use_yamlloader_old`](https://docs.saltproject.io/en/latest/ref/configuration/master.html#use-yamlloader-old).  That option doesn't appear to be marked as deprecated.  Is it old enough that we can safely remove it to avoid confusion with a new option?

I'm tempted to make the option a version selector rather than a boolean so that the YAML behavior can be changed again in the future without much fanfare.  Maybe something like `override_yaml_parser_version` that defaults to the current version of Salt and behaves like this:

```python
from salt.version import SaltStackVersion

want_v = __opts__.get("override_yaml_parser_version")
if want_v:
    want_v = SaltStackVersion.parse(want_v)
else:
    want_v = SaltStackVersion.current_release()

if want_v <= SaltStackVersion.parse("3005.1"):
    do_old_stuff()
else:
    do_this_new_stuff()
```

Note that the new behavior would be the default behavior, so if it breaks a user they would have to take action to switch to the old behavior.

Thoughts?

**Comment by rhansen** [comment]
> Note that the new behavior would be the default behavior, so if it breaks a user they would have to take action to switch to the old behavior.

If we think this change (or a future change) is too risky to make it the default right away, we could predate the version and add a warning:

```python
if want_v < SaltStackVersion.parse("3007.0"):
    if not __opts__.get("override_yaml_parser_version"):
        salt.utils.versions.warn_until(3007, "Salt's YAML parsing behavior will change in v3007.0; see xyz for details")
    do_old_stuff()
else:
    do_this_new_stuff()
```

File: salt/utils/yamlloader.py [hunk 3]
```diff
@@ -24,27 +26,46 @@ class SaltYamlSafeLoader(BaseLoader):
...
+        for k, v in entries:
+            if k in d:
+                raise ConstructorError(
+                    f"while constructing an ordered map",
+                    node.start_mark,
+                    f"duplicate key encountered: {k!r}",
+                    # TODO: Can we get the location of the duplicate key?
+                    node.start_mark,
+                )
+            d[k] = v
```
Comment (whytewolf) [inline salt/utils/yamlloader.py:67]:
```suggestion
        inter_d_entries = set(entries).intersection(d)
        if len(inter_d_entries) > 0:
            raise ConstructorError(
                f"while constructing an ordered map",
                node.start_mark,
                f"duplicate keys encountered: {inter_d_entries}",
                node.start_mark,
            )
        d.update(entries)
```

don't need the loop and can report all of the duplicates in a single exception.

Comment (rhansen) [reply salt/utils/yamlloader.py:67]: The goal is to identify duplicates in `entries`, not to identify overlap between `entries` and `d`.  Note that `d` is empty at the time `set(entries).intersection(d)` would be called, so the condition would always be false.

**Comment by rhansen** [comment]
re-run full all

**Comment by rhansen** [comment]
I set this to draft until I do the following:
  * [ ] Add a feature flag to enable this new behavior.
  * [ ] Make sure the full set of tests pass.

**Comment by rhansen** [comment]
@waynew I'm trying to figure out how to coax the loader to set `__opts__` in `salt.utils.yaml` and friends, but the loader documentation is wafer thin.  Any hints?

**Comment by MKLeb** [comment]
@rhansen There is an example of how to mock the loader dunders [here](https://github.com/saltstack/salt/blob/932d4c19810c26f94b5a1729130f5d459db6831e/tests/pytests/unit/utils/test_vault.py#L28)

**Comment by rhansen** [comment]
@MKLeb Thanks, though it looks like `__opts__` isn't available in `salt.utils.yaml` when Salt normally runs (not in test) because `salt.utils.yaml` is usually `import`ed, not loaded through the loader.  Is that correct?  Can I manually trigger the loader at daemon startup to populate the dunder dictionaries so that `__opts__` is available even if `salt.utils.yaml` is directed `import`ed?

**Comment by MKLeb** [comment]
@s0undt3ch Any ideas here?

**Comment by lkubb** [comment]
Not sure if that's helpful, but you can always manually load `__opts__` and friends (watch out for performance though): https://github.com/saltstack/salt/blob/327664a375fdd133b1d347bdc15726af5883fab6/salt/loader/__init__.py#L291-L303

```python
import salt.config

__opts__ = None


def my_func():
    global __opts__
    if __opts__ is None:
        __opts__ = salt.config.minion_config('/etc/salt/minion')
    # in case you need grains in __opts__, see the linked example
    # ... do stuff
```

**Comment by rhansen** [comment]
My thought was to do this:

```python
import salt.loader.context

__salt_loader__ = salt.loader.context.LoaderContext()
__opts__ = __salt_loader__.named_context("__opts__", {})
```

Unfortunately that didn't work due to a circular dependency:
  1. `saltfactories` does `import salt.utils.user`
  2. which does `import salt.utils.path`
  3. which does `import salt.utils.args`
  4. which does `import salt.utils.data`
  5. which does `import salt.utils.yaml`
  6. which does `import salt.loader.context`
  7. which does `import salt.config` (via `salt.loader`)
  8. which does `import salt.utils.user`

So now I defer the `import` until `__opts__` is actually needed.  That fixed that problem, but a new problem has surfaced:  I get a `KeyError: '__opts__'` exception when I try to read the `__opts__` mapping object:

```
Traceback (most recent call last):
  File "salt/pillar/__init__.py", line 917, in render_pstate
    state = compile_template(
  File "salt/template.py", line 99, in compile_template
    ret = render(input_data, saltenv, sls, **render_kwargs)
  File "salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
  File "salt/renderers/yaml.py", line 63, in render
    data = yamlloader.load(yaml_data, Loader=get_yaml_loader(argline))
  File "salt/utils/yamlloader.py", line 206, in load
    return yaml.load(stream, Loader=Loader)
  File ".nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.8/site-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File ".nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File ".nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.8/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "salt/utils/yamlloader.py", line 46, in construct_yaml_map
    value = self.construct_mapping(node)
  File "salt/utils/yamlloader.py", line 103, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
  File ".nox/test-parametrized-3-crypto-none-transport-zeromq-coverage-false/lib/python3.8/site-packages/yaml/constructor.py", line 105, in construct_object
    data = next(generator)
  File "salt/utils/yamlloader.py", line 52, in construct_yaml_omap
    if compat_ver() < SaltStackVersion(3007):
  File "salt/utils/_yaml_compat.py", line 44, in compat_ver
    opt = __opts__.get("yaml_compatibility")
  File "salt/loader/context.py", line 75, in get
    return self.value().get(key, default)
  File "salt/loader/context.py", line 72, in value
    return loader.pack[self.name]
KeyError: '__opts__'
```

I feel like I'm on the right track, but the loader's complexity is making it hard for me to figure out what lever I need to pull.

> you can always manually load `__opts__` and friends (watch out for performance though):

Thanks, although I'm worried that that approach is not the Right Way, and will therefore be fragile.  In particular, using the minion path like that won't work for the master (I think?), which will affect pillar rendering.  It also won't work for the `salt` CLI (I think?), which will affect `--out yaml`.

**Comment by rhansen** [comment]
Success!  If I modify `salt/loader/__init__.py` to add `"__opts__": opts` to each of the `LazyLoader` constructor `pack` dicts, then the `KeyError` goes away and I get the value from the config.  I'll polish it up and push a new revision tomorrow.

**Comment by rhansen** [comment]
Using a feature flag is turning out to be trickier than expected due to circular imports.  I have a solution involving lazily loaded imports that I put in a different pull request (see https://github.com/saltstack/salt/pull/63073).  Once that is merged I'll rebase this PR.

**Comment by rhansen** [comment]
re-run full all

**Comment by dwoz** [comment]
I am very much not sure how I feel about this change. The lazy loading stuff is pretty concerning. I feel like there should be a more direct (hopefully simpler) way to achieve the performance improvements we are after here.

**Comment by whytewolf** [comment]
honestly. the more commits I see to this the less i like it. this is currently at SEP levels of change for me.

**Comment by rhansen** [comment]
> The lazy loading stuff is pretty concerning.

That's why I put it in a separate pull request:  #63073 

> the more commits I see to this the less i like it.

Most of the complexity can be removed (including the lazy load stuff) if we don't care about potential compatibility issues.  I added it in response to [review feedback](https://github.com/saltstack/salt/pull/62932#issuecomment-1292586738).  My preference would be to eliminate the backwards compatibility logic and hope it doesn't break too many users.

**Comment by rhansen** [comment]
I'm hoping the documentation, test, and readability improvement commits are non-controversial, so I moved them to their own PR: #63158

**Comment by whytewolf** [comment]
> > the more commits I see to this the less i like it.
> 
> Most of the complexity can be removed (including the lazy load stuff) if we don't care about potential compatibility issues. I added it in response to [review feedback](https://github.com/saltstack/salt/pull/62932#issuecomment-1292586738). My preference would be to eliminate the backwards compatibility logic and hope the behavior changes don't break too many users.

backwords compatibility is not optional. yaml is to much of a core item for backwords compatibility to be a hope and a prayer. 

what is concerning is the magic importing wrapping esp considering you are wrapping the loader which is already a magic import system in another magic import system. your lazy loading at a bare min needs to go through a SEP. see https://github.com/saltstack/salt-enhancement-proposals

**Comment by rhansen** [comment]
Moved discussion of `lazy_import` to https://github.com/saltstack/salt/pull/63073#issuecomment-1334538891

**Comment by dwoz** [comment]
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.

**Pull request closed without merging**